pkg/virt-handler/vm.go

```diff
@@ -2909,15 +2923,24 @@ func (d *VirtualMachineController) deleteDomainFunc(obj interface{}) {
 func (d *VirtualMachineController) updateDomainFunc(old, new interface{}) {
 	newDomain := new.(*api.Domain)
 	oldDomain := old.(*api.Domain)
+	key, err := controller.KeyFunc(new)
 	if oldDomain.Status.Status != newDomain.Status.Status || oldDomain.Status.Reason != newDomain.Status.Reason {
 		log.Log.Object(newDomain).Infof("Domain is in state %s reason %s", newDomain.Status.Status, newDomain.Status.Reason)
+		if oldDomain.Status.Status == api.Running && newDomain.Status.Status == api.Paused && newDomain.Status.Reason == api.ReasonPausedIOError {
+			if !d.backoffHandler.IsInBackOffSinceUpdate(key, time.Now()) {
+				d.backoffHandler.Next(key, time.Now())
+				log.Log.Object(newDomain).Infof("Delaying re-enqueue due to I/O Error. Backoff time is %v", d.backoffHandler.Get(key))
+				d.Queue.AddAfter(key, d.backoffHandler.Get(key))
+				return
+			}
```

Comment:
I found it hard for me to understand the logic here. When the event first happens, the key is not in backoff, you will put it in backoff, and requeue it, which means the controller will not trigger sync and thus unpause for this domain update. Then when the requeued event came after the backoff time, since it came after the backoff delay time,  it will be again not in backoff.  Aren't you going to requeue the request again? Or you are relying on something else so that the requeue will come early and thus pass the check and continue? Even it passed, it is still in backoff and because your changes in line 1726,  it will not be processed either. Then when will the unpause be triggered? Am I missing something? I cannot tell the logic from the unit test example either. 

Ideally we want the logic to happen on a single place instead of multiple places, and avoid relying on too much on the underlying process logic. May we consider putting all retry logic into defaultExecute function instead?  We have an internal implementation that fix this issue, and will be happy to share and consolidate the fix for OSS together.

Reply:
Thanks @lzang for your review.
Here's the logic:
1) Domain fails due to IOError
1) The domain informer receive the update event and since the domains goes to Paused due to IOError increases the backoff and re-enqueue after the backoff time (Event t).
1) When Event t comes the Execute function handles the request (it goes through `shouldUpdate = true` since the backoff expires), and the vmi will be unpaused
1) If Another IOError occurs return to point 1)
I understand what are you saying; but AFAIK the updateDomainFunction is the only place where I can see the old domain and the new one, so that we can delay the re-enqueue only when the domain goes from Running to Paused.
In the defaultExecute we only have the new domain.
I'll be happy if you are able to share your implementation and work together to address it. :)

Reply:
Thanks @fossedihelm for your explanation! What confused me is how step2 transits to step 3 because step2 seems to always re-enqueue instead of getting out unless line 2929 is false when you re-enqueue. Let's sync offline on this - I will ping you. Thanks!

Reply:
The `d.Queue.AddAfter(key, d.backoffHandler.Get(key))` puts the key in the queue after the backoff time.
This queue is read from the VirtualMachineController during Execute function, which handles the vm-cycle.
https://github.com/kubevirt/kubevirt/blob/b4d29db49651e350286463ff0856ed94052dc2bd/pkg/virt-handler/vm.go#L1433
In step2 we only react to a domain change. Currently, every domain changes trigger the adding of the key immediately, which means that the `Execute` will be executed. This patch just add a condition: if it is paused due to an I/O error do not execute immediately, but only after a backoff time. 
Hope this helps to clarify, thanks!

Reply:
Ah, my bad. I mistook it as re-queue of itself instead of enqueue. Now I got your logic :-) Thanks for the clarification.